src/Compilers/Core/Portable/SourceGeneration/CachingSyntaxTreeProvider.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class CachingSyntaxTreeProvider : SyntaxTreeProvider
+    {
+        public static readonly CachingSyntaxTreeProvider Instance = new();
+
+        private static readonly ConditionalWeakTable<SourceText, SyntaxTree> s_syntaxTrees = new();
+
+        private CachingSyntaxTreeProvider()
+        {
+        }
+
+        public override bool TryGetSyntaxTree(SourceText sourceText, [NotNullWhen(true)] out SyntaxTree? tree)
+        {
+            return s_syntaxTrees.TryGetValue(sourceText, out tree);
+        }
+
+        public override void AddOrUpdate(SourceText text, SyntaxTree tree)
```

Comment:
Where do we explicitly clear this cache/CWT? For semantic model provider used by the analyzer driver, we have: https://github.com/dotnet/roslyn/blob/6352ca4d53b24b35cf57f21584040b2c5d4e34b5/src/Compilers/Core/Portable/DiagnosticAnalyzer/CachingSemanticModelProvider.cs#L37-L48

There is a clear lifetime when all compilation events have finished processing for each compilation unit, and so we drop the cached semantic model, which provides much stronger control over the cache size. Unless we can do something similar for this caching provider, this seems no better then just using a CWT directly in the generator driver.

Reply:
> Where do we explicitly clear this cache/CWT? 

We don't, which is the basis for this cache working.

Reply:
I don't see how a provider based approach helps for your scenario then. There are dual purposes of semantic model provider being hooked up to compilation for analyzer execution - one is that the public API Compilation.GetSemanticModel shares semantic models (so all analyzers share models), other that the analyzer driver itself can clear the cache when events for each tree are fully processed. If there is just a single client for the cache and no way to clear cache (which seems to be your scenario), then using a CWT directly inside generator driver seems better than having this intermediate provider.

Reply:
> I don't see how a provider based approach helps for your scenario then

It addresses the blocking change request in https://github.com/dotnet/roslyn/pull/50171#discussion_r551396142. I found the result to be a bit less readable than before but hopefully we can use it as a step towards a conclusion.

Reply:
Stating more empirically, https://github.com/dotnet/roslyn/pull/46508 that introduced semantic model provider showed a clear performance win for various analyzer execution benchmarks. I believe any approach taken by this PR needs to demonstrate similar performance wins to justify it with current approach or any alternate approaches.

Reply:
> It addresses the blocking change request in #50171 (comment).

I believe @jaredpar's request was not purely in terms of changing just the implementation to use a provider, but actually mimicing the stronger lifetime control over the cache that is done is analyzer driver's semantic model provider, which optimized cache size and gave real performance improvements over a purely CWT based approach. Your change seems to just introduce an artificial provider, which uses a CWT under the covers but without any of the above benefits, so I am not sure if addresses the core concern.

Reply:
For me the issue is about policy and who is controlling it. 

From the perspective of batch compilation this cache isn't really needed. The addition of the cache is being driven by IDE needs. The IDE is in the best place to dictate what the lifetime of this cache is, when it should be evicted, etc ... 

Pretty much every time we've put a CWT in the compiler layer to do a cache for IDE purposes we seem to regret it. Instead it seems better to structure it such that the IDE can be in control and make better decisions about their caches. That is why I had my recommendation about looking at the provider approach that we've taken elsewhere in the code.

Reply:
> the stronger lifetime control over the cache that is done is analyzer driver's semantic model provider

I'm not sure that scenario is relevant to this change. Semantic models never need to carry forward from one `Compilation` to another, while the entire value of this pull request is allowing `SyntaxTree` implementations to carry forward. The only good time to remove an item from the cache is when the SG is no longer going to provide the same `SourceText` instance.

Reply:
To restate: this is an IDE cache we're adding into the compiler layer. The IDE has no direct, or even indirect, control over the lifetime or semantics of it. From a conceptual level that feels wrong. All the bugs here (cache too big, elements live too long, etc ...) will be IDE bugs and we'll end up fixing everything in the compiler. 

That is why my mind went back to the semantic model provider solution. It's putting the control, the policy, etc ... for an IDE cache in the hands of the IDE team. 

I'm not 100% set on this on a path forward but my feelings around compiler dictating IDE caches keep coming up.

Reply:
My understanding of the plan now is we use this simple approach to mitigate the performance problems in 16.9, and then work to replace it with https://github.com/dotnet/roslyn/pull/50490 for 16.10.